Warn if "include directories" do not exist

josephsnyder opened the pull request:
Instead of raising a RunTime error on a missing include
directory, print a warning to the screen and allow processing to continue.

tao558 commented on pygccxml/parser/config.py:
```diff
@@ -200,8 +200,12 @@ def __ensure_dir_exists(self, dir_path, meaning):
-            raise RuntimeError(
-                '%s("%s") does not exist!' % (meaning, dir_path))
+            msg = '%s("%s") does not exist!' % (meaning, dir_path)
+            if meaning == 'include_directory':
```
```suggestion
            if meaning == 'include directory':
```

EricCousineau-TRI commented on unittests/test_warn_missing_include_dirs.py:
```diff
@@ -0,0 +1,58 @@
+
+        f = io.StringIO()
+        # Path given as include director doesn't exist
+        config = parser.xml_generator_configuration_t(
```
Per f2f, can we include a mechanism to pass a `-Werror` proxy, so that this warning is still an error?

Context:
In https://github.com/RobotLocomotion/drake/pull/14265, we actually do want to ensure those directories exist (I think).

EricCousineau-TRI replied:
OK I think this was resolved by a separate mechanism. I dunno how to resolve my own review comment, though.

iMichka reviewed (changes requested):
I think we should use the warnings module to emit the warning instead of printing it:
https://docs.python.org/3/library/warnings.html

This allows a finer filtering for the users.

I think we even have some code to test warnings here: https://github.com/CastXML/pygccxml/blob/1ed25bda6a25f7f1bd4efdf9e6ed64c43afe592f/unittests/test_utils.py#L43-L44

iMichka commented:
Thinking even more about it, now that we use the warning module, one can use (as a user):
```
import warnings
warnings.filterwarnings("error")
```

This allows to transform warnings into errors. And with that we do not need the whole error handling flags.

The warning code needs to have a warning type so that people can filter on specific warnings:

```
warnings.warn(msg, RuntimeWarning)
```

Sorry for the back and forth. We already that concept elsewhere:
https://github.com/CastXML/pygccxml/blob/2b1efbb9e37ceb2ae925c7f3ce1570f476db9e1e/pygccxml/utils/utils.py#L174-L179

EricCousineau-TRI commented:
FWIW I'd recommend having project-specific warnings, just make it a tad easier.
We do that in Drake here:
https://github.com/RobotLocomotion/drake/blob/5b5d6aa8efbd57cbf1fc7b1fdd95d78210290668/bindings/pydrake/common/deprecation.py#L109-L112
It allows simpler deprecation, and can add a layer of assurance that you're checking the right project, vs. crossing your fingers that the error message has sufficient info to be unique, e.g.:
https://github.com/RobotLocomotion/drake/blob/5b5d6aa8efbd57cbf1fc7b1fdd95d78210290668/bindings/pydrake/common/deprecation.py#L199-L221

iMichka commented:
Regarding the project specific warnings, it's a good idea. More projects should do that.

I would recommend doing this in a second PR, in a more global way, for the whole pygccxml project. Let's try to finish this change without adding more feature creep in this pull request.

josephsnyder commented:
Thank you both for the comments.  I've updated the branch to eliminate the new flag commit but have kept the commit locally for additional updates when we think it is time.  I've also added the warning type and updated the test to account for it.

EricCousineau-TRI commented:
Sounds good!

tao558 commented:
Is everyone all set with this PR? @iMichka @EricCousineau-TRI
